tracing/span.go

```diff
@@ -209,91 +162,162 @@ func (s *Span) ToZipkinSpan() ZipkinSpan {
 	return zs
 }
 
-func (s *Span) isDebugSet() bool {
-	return s.flags&FlagMaskDebug != 0
+func (t *trace) isDebugSet() bool {
+	return t.flags&FlagMaskDebug != 0
 }
 
-// ShouldSample returns true if this span should be sampled.
+func (t *trace) setDebug(v bool) {
+	if v {
+		t.flags |= FlagMaskDebug
+	} else {
+		t.flags &= ^FlagMaskDebug
+	}
+}
+
+// shouldSample returns true if this span should be sampled.
 //
 // If the span's debug flag was set, then this function will always return true.
-func (s *Span) ShouldSample() bool {
-	return s.sampled || s.isDebugSet()
+func (t *trace) shouldSample() bool {
+	return t.sampled || t.isDebugSet()
 }
 
-// End ends a span.
-//
-// If err is non-nil, the error tag will also be set to true.
-//
-// It also sends the span into tracer, if the span should be sampled.
-// The context object passed in is used to control the timeout of the sending,
-// see Tracer.MaxRecordTimeout for more info.
-func (s *Span) End(ctx context.Context, err error) error {
-	s.end = time.Now()
+func (t *trace) stopAndPublish(ctx context.Context, err error) error {
+	t.stop = time.Now()
 
-	if err != nil {
-		s.SetTag(ZipkinBinaryAnnotationKeyError, true)
-		if s.spanType == SpanTypeServer && errors.Is(err, context.DeadlineExceeded) {
-			s.SetTag(ZipkinBinaryAnnotationKeyTimeOut, true)
-		}
+	if !t.shouldSample() || t.tracer == nil {
+		return nil
 	}
-	if s.isDebugSet() {
-		s.SetTag(ZipkinBinaryAnnotationKeyDebug, true)
+
+	if err != nil {
+		t.setTag(ZipkinBinaryAnnotationKeyError, true)
 	}
-	if s.spanType == SpanTypeLocal && s.component != "" {
-		s.SetTag(ZipkinBinaryAnnotationKeyLocalComponent, s.component)
+
+	if t.isDebugSet() {
+		t.setTag(ZipkinBinaryAnnotationKeyDebug, true)
 	}
-	for _, hook := range s.hooks {
-		if hookErr := hook.OnEnd(s, err); hookErr != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnEnd hook error: " + hookErr.Error())
+
+	return t.tracer.Record(ctx, t.toZipkinSpan())
+}
+
+// Span interface
+type Span interface {
+	Name() string
+	SpanType() SpanType
+	Hooks() []SpanHook
+	AddHooks(hooks ...SpanHook)
+	Start()
+	Stop(err error)
+	Publish(ctx context.Context, err error) error
+	SetTag(key string, value interface{})
+	AddCounter(key string, delta float64)
+	CreateLocalChild(name, component string) *LocalSpan
+	CreateClientChild(name string) *ClientSpan
+	LogError(msg string, err error)
+	SetDebug(v bool)
+}
```

Comment:
Same here, please add comments to the functions, especially for functions not obvious, e.g. `Stop` vs. `Publish`.